src/shared/mkfs-util.c

```diff
@@ -185,6 +188,96 @@ static int do_mcopy(const char *node, const char *root) {
         return 0;
 }
 
+static int protofile_print_item(
+                RecurseDirEvent event,
+                const char *path,
+                int dir_fd,
+                int inode_fd,
+                const struct dirent *de,
+                const struct statx *sx,
+                void *userdata) {
+
+        FILE *f = ASSERT_PTR(userdata);
+        _cleanup_free_ char *base = NULL;
+        int r;
+
+        if (event == RECURSE_DIR_LEAVE) {
+                fprintf(f, "$\n");
+                return 0;
+        }
+
+        if (!IN_SET(event, RECURSE_DIR_ENTER, RECURSE_DIR_ENTRY))
+                return RECURSE_DIR_CONTINUE;
+
+        r = path_extract_filename(path, &base);
+        if (r < 0)
+                return log_error_errno(r, "Failed to extract basename from %p: %m", path);
```

Comment:
just use `de->d_name`. no need to extract from the path name what was appended there just instants ago, if you can just get it directly.